src/audio/detect_test.c

```diff
@@ -0,0 +1,224 @@
+
+struct comp_data {
+	uint32_t period_bytes;
+	uint32_t switch_state[2];
+};
+
+static struct comp_dev *keyword_new(struct sof_ipc_comp *comp)
+{
+	struct comp_dev *dev;
+	struct sof_ipc_comp_process *keyword;
+	struct sof_ipc_comp_process *ipc_keyword =
+		(struct sof_ipc_comp_process *)comp;
+	struct comp_data *kd;
+
+	trace_keyword("keyword_new()");
+
+	if (IPC_IS_SIZE_INVALID(ipc_keyword->config)) {
+		IPC_SIZE_ERROR_TRACE(TRACE_CLASS_MUX, ipc_keyword->config);
+		return NULL;
+	}
+
+	dev = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM,
+		COMP_SIZE(struct sof_ipc_comp_process));
+	if (!dev)
+		return NULL;
+
+	keyword = (struct sof_ipc_comp_process *)&dev->comp;
+	memcpy(keyword, ipc_keyword, sizeof(struct sof_ipc_comp_process));
+
+	kd = rzalloc(RZONE_RUNTIME, SOF_MEM_CAPS_RAM, sizeof(*kd));
+	if (!kd) {
+		rfree(dev);
+		return NULL;
+	}
+
+	comp_set_drvdata(dev, kd);
+	dev->state = COMP_STATE_READY;
+	return dev;
+}
+
+static void keyword_free(struct comp_dev *dev)
+{
+	struct keyword_data *kd = comp_get_drvdata(dev);
+
+	trace_keyword("keyword_free()");
+
+	rfree(kd);
+	rfree(dev);
+}
+
+/* set component audio stream parameters */
+static int keyword_params(struct comp_dev *dev)
+{
+
+	return 0;
+}
+
+/**
+ * \brief Sets keyword control command.
+ * \param[in,out] dev keyword base component device.
+ * \param[in,out] cdata Control command data.
+ * \return Error code.
+ */
+static int keyword_ctrl_set_cmd(struct comp_dev *dev,
+			       struct sof_ipc_ctrl_data *cdata)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int j;
+
+	/* validate */
+	if (cdata->num_elems == 0 || cdata->num_elems > SOF_IPC_MAX_CHANNELS) {
+		trace_keyword_error("keyworkd_ctrl_set_cmd() error: "
+				   "invalid cdata->num_elems");
+		return -EINVAL;
+	}
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_SWITCH:
+		/*FIXME: fix logger for keywork */
+		trace_keyword("keyword_ctrl_set_cmd(), SOF_CTRL_CMD_SWITCH, ");
+
+		/* save switch state */
+		for (j = 0; j < cdata->num_elems; j++)
+			cd->switch_state[j] = cdata->chanv[j].value;
+
+		break;
+
+	default:
+		trace_keyword_error("keyword_ctrl_set_cmd() error: "
+				   "invalid cdata->cmd");
+		return -EINVAL;
+	}
+
+	return 0;
+}
+
+/**
+ * \brief Gets keyword control command.
+ * \param[in,out] dev keyword base component device.
+ * \param[in,out] cdata Control command data.
+ * \return Error code.
+ */
+static int keyword_ctrl_get_cmd(struct comp_dev *dev,
+			       struct sof_ipc_ctrl_data *cdata, int size)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int j;
+
+	/* validate */
+	if (cdata->num_elems == 0 || cdata->num_elems > SOF_IPC_MAX_CHANNELS) {
+		trace_keyword_error("invalid cdata->num_elems");
+		return -EINVAL;
+	}
+
+	if (cdata->cmd ==  SOF_CTRL_CMD_SWITCH) {
+		trace_keyword("keyword_ctrl_get_cmd(), SOF_CTRL_CMD_SWITCH");
+
+		/* copy current switch state */
+		for (j = 0; j < cdata->num_elems; j++) {
+			cdata->chanv[j].channel = j;
+			cdata->chanv[j].value = cd->switch_state[j];
```

Comment:
There is an issue detected in this line: `switch_state` size is h/c to 2 while `cdata->enum_elems` is validated against `SOF_IPC_MAX_CHANNELS` (8) in L152.